File: 24_07_2024_Check_for_BST.py
class Solution:
    def __init__(self):
        self.prev=None
    
    #Function to check whether a Binary Tree is BST or not.
    def isBST(self, root):
        return self.inorder(root)
    
    def inorder(self, root):
        if root is None:
            return True
        
        if not self.inorder(root.left):
            return False
        
        # Check BST condition
        if self.prev is not None and root.data <= self.prev:
            return False
        
        self.prev = root.data # update prev to current node's value
        
        return self.inorder(root.right)

File: 24_07_2024_Check_for_BST.java
import java.util.Objects;

class Solution {
    private Integer prev;

    // Function to check whether a Binary Tree is BST or not.
    public boolean isBST(TreeNode root) {
        return inorder(root);
    }

    private boolean inorder(TreeNode root) {
        if (root == null) {
            return true;
        }

        if (!inorder(root.left)) {
            return false;
        }

        // Check BST condition
        if (prev != null && root.data <= prev) {
            return false;
        }

        prev = root.data; // update prev to current node's value

        return inorder(root.right);
    }
}

class TreeNode {
    int data;
    TreeNode left, right;

    TreeNode(int item) {
        data = item;
        left = right = null;
    }
}

